File: src/main/kotlin/com/babasama/edendale/wyzie/WyzieSubtitleService.kt
package com.babasama.edendale.wyzie

import java.net.URI
import java.net.URLEncoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive

data class WyzieSubtitle(
    val id: String,
    val url: String,
    val format: String,
    val encoding: String,
    val isHearingImpaired: Boolean,
    val flagUrl: String,
    val media: String,
    val display: String,
    val language: String,
    val source: List<String>? = null,
    val release: String? = null,
    val releases: List<String>? = null,
    val fileName: String? = null,
    val downloadCount: Int? = null,
    val origin: String? = null,
    val matchedRelease: String? = null,
    val matchedFilter: String? = null,
)

data class WyzieSubtitleQuery(
    val id: String,
    val season: Int? = null,
    val episode: Int? = null,
    val language: String? = null,
    val format: String? = null,
    val hearingImpaired: Boolean? = null,
)

sealed class WyzieException(message: String) : Exception(message) {
    class MissingKey : WyzieException("A Wyzie API key is required.")

    class BadStatus(
        val code: Int,
        val serverMessage: String?,
    ) : WyzieException("Wyzie request failed (HTTP $code).")

    class EmptyFile : WyzieException("The downloaded subtitle was empty.")
    class FileTooLarge : WyzieException("The downloaded subtitle exceeded the size limit.")
    class BadUrl : WyzieException("The subtitle URL was not an HTTP or HTTPS URL.")
}

data class WyzieResponse(
    val statusCode: Int,
    val body: String,
)

interface WyzieTransport {
    suspend fun get(url: String): WyzieResponse
    suspend fun download(url: String): ByteArray
}

fun queryParameters(
    query: WyzieSubtitleQuery,
    key: String,
): List<Pair<String, String>> = buildList {
    add("id" to query.id)
    query.season?.let { season ->
        add("season" to season.toString())
        query.episode?.let { add("episode" to it.toString()) }
    }
    query.language?.let { add("language" to it) }
    query.format?.let { add("format" to it) }
    query.hearingImpaired?.let { add("hi" to it.toString()) }
    add("key" to key)
}

fun searchUrl(query: WyzieSubtitleQuery, key: String): String =
    "$BASE_URL/search?" + queryParameters(query, key).joinToString("&") { (name, value) ->
        "${name.percentEncoded()}=${value.percentEncoded()}"
    }

fun decodeSearchResponse(body: String, statusCode: Int): List<WyzieSubtitle> {
    if (statusCode !in 200..299) {
        throw badStatus(statusCode, body)
    }
    val payload = Json.parseToJsonElement(body)
    val rows = payload as? JsonArray
        ?: throw IllegalArgumentException("Wyzie success response was not an array.")
    return rows.mapNotNull { element ->
        val item = element as? JsonObject ?: return@mapNotNull null
        WyzieSubtitle(
            id = item.string("id") ?: return@mapNotNull null,
            url = item.string("url") ?: return@mapNotNull null,
            format = item.string("format") ?: return@mapNotNull null,
            encoding = item.string("encoding") ?: return@mapNotNull null,
            isHearingImpaired = item.boolean("isHearingImpaired") ?: return@mapNotNull null,
            flagUrl = item.string("flagUrl") ?: return@mapNotNull null,
            media = item.string("media") ?: return@mapNotNull null,
            display = item.string("display") ?: return@mapNotNull null,
            language = item.string("language") ?: return@mapNotNull null,
            source = item.stringOrArray("source"),
            release = item.string("release"),
            releases = item.stringArray("releases"),
            fileName = item.string("fileName"),
            downloadCount = item.int("downloadCount"),
            origin = item.string("origin"),
            matchedRelease = item.string("matchedRelease"),
            matchedFilter = item.string("matchedFilter"),
        )
    }
}

fun cacheFileName(subtitle: WyzieSubtitle): String {
    val id = subtitle.id.safeFilePart()
    val language = subtitle.language.safeFilePart()
    val format = subtitle.format.safeFilePart().ifBlank { "srt" }
    return "wyzie-$id-$language.$format"
}

class WyzieSubtitleService(private val transport: WyzieTransport) {
    suspend fun search(query: WyzieSubtitleQuery, key: String): List<WyzieSubtitle> {
        val resolvedKey = key.trim()
        if (resolvedKey.isEmpty()) throw WyzieException.MissingKey()
        val response = transport.get(searchUrl(query, resolvedKey))
        return decodeSearchResponse(response.body, response.statusCode)
    }

    suspend fun download(subtitle: WyzieSubtitle): ByteArray {
        val uri = runCatching { URI(subtitle.url) }.getOrNull()
        if (uri?.scheme?.lowercase() !in setOf("http", "https") || uri?.host.isNullOrBlank()) {
            throw WyzieException.BadUrl()
        }
        val body = transport.download(subtitle.url)
        if (body.isEmpty()) throw WyzieException.EmptyFile()
        if (body.size > MAX_SUBTITLE_BYTES) throw WyzieException.FileTooLarge()
        return body
    }
}

internal fun badStatus(statusCode: Int, body: String): WyzieException.BadStatus {
    val message = runCatching {
        (Json.parseToJsonElement(body) as? JsonObject)?.string("message")
    }.getOrNull()
    return WyzieException.BadStatus(statusCode, message)
}

private fun JsonObject.string(key: String): String? =
    get(key)?.jsonPrimitive?.contentOrNull

private fun JsonObject.int(key: String): Int? =
    get(key)?.jsonPrimitive?.intOrNull

private fun JsonObject.boolean(key: String): Boolean? =
    get(key)?.jsonPrimitive?.booleanOrNull

private fun JsonObject.stringArray(key: String): List<String>? =
    (get(key) as? JsonArray)
        ?.mapNotNull { it.jsonPrimitive.contentOrNull }

private fun JsonObject.stringOrArray(key: String): List<String>? = when (val value = get(key)) {
    is JsonArray -> value.mapNotNull { it.jsonPrimitive.contentOrNull }
    null -> null
    else -> value.jsonPrimitive.contentOrNull?.let(::listOf)
}

private fun String.safeFilePart(): String =
    lowercase().filter { it in 'a'..'z' || it in '0'..'9' || it == '-' || it == '_' }

private fun String.percentEncoded(): String =
    URLEncoder.encode(this, Charsets.UTF_8.name()).replace("+", "%20")

private const val BASE_URL = "https://sub.wyzie.io"
internal const val MAX_SUBTITLE_BYTES = 5 * 1024 * 1024

File: Shared/Controllers/WyzieSubtitleService.swift
//
//  WyzieSubtitleService.swift
//  Edendale
//
//  Online subtitle search through sub.wyzie.io. A synchronized user key takes
//  precedence over the optional build-time key, searches happen only after an
//  explicit user action, and validated downloads are stored only in Caches.
//

import Foundation

struct WyzieSubtitle: Identifiable, Sendable, Decodable, Hashable {
    let id: String
    let url: String
    let format: String
    let encoding: String
    let isHearingImpaired: Bool
    let flagUrl: String
    let media: String
    let display: String
    let language: String
    let source: [String]?
    let release: String?
    let releases: [String]?
    let fileName: String?
    let downloadCount: Int?
    let origin: String?
    let matchedRelease: String?
    let matchedFilter: String?

    init(
        id: String,
        url: String,
        format: String,
        encoding: String,
        isHearingImpaired: Bool,
        flagUrl: String,
        media: String,
        display: String,
        language: String,
        source: [String]? = nil,
        release: String? = nil,
        releases: [String]? = nil,
        fileName: String? = nil,
        downloadCount: Int? = nil,
        origin: String? = nil,
        matchedRelease: String? = nil,
        matchedFilter: String? = nil
    ) {
        self.id = id
        self.url = url
        self.format = format
        self.encoding = encoding
        self.isHearingImpaired = isHearingImpaired
        self.flagUrl = flagUrl
        self.media = media
        self.display = display
        self.language = language
        self.source = source
        self.release = release
        self.releases = releases
        self.fileName = fileName
        self.downloadCount = downloadCount
        self.origin = origin
        self.matchedRelease = matchedRelease
        self.matchedFilter = matchedFilter
    }

    private enum CodingKeys: String, CodingKey {
        case id, url, format, encoding, isHearingImpaired, flagUrl, media
        case display, language, source, release, releases, fileName
        case downloadCount, origin, matchedRelease, matchedFilter
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        url = try container.decode(String.self, forKey: .url)
        format = try container.decode(String.self, forKey: .format)
        encoding = try container.decode(String.self, forKey: .encoding)
        isHearingImpaired = try container.decode(Bool.self, forKey: .isHearingImpaired)
        flagUrl = try container.decode(String.self, forKey: .flagUrl)
        media = try container.decode(String.self, forKey: .media)
        display = try container.decode(String.self, forKey: .display)
        language = try container.decode(String.self, forKey: .language)
        source = try container.decodeIfPresent(StringOrArray.self, forKey: .source)?.values
        release = try container.decodeIfPresent(String.self, forKey: .release)
        releases = try container.decodeIfPresent([String].self, forKey: .releases)
        fileName = try container.decodeIfPresent(String.self, forKey: .fileName)
        downloadCount = try container.decodeIfPresent(Int.self, forKey: .downloadCount)
        origin = try container.decodeIfPresent(String.self, forKey: .origin)
        matchedRelease = try container.decodeIfPresent(String.self, forKey: .matchedRelease)
        matchedFilter = try container.decodeIfPresent(String.self, forKey: .matchedFilter)
    }
}

private struct StringOrArray: Decodable {
    let values: [String]

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode(String.self) {
            values = [value]
        } else {
            values = try container.decode([String].self)
        }
    }
}

enum WyzieError: Error, LocalizedError {
    case missingKey
    case badStatus(code: Int, message: String?)
    case emptyFile
    case fileTooLarge
    case badURL

    var errorDescription: String? {
        switch self {
        case .missingKey:
            String(localized: "A Wyzie API key is required.")
        case .badStatus(let code, let message):
            if let message, !message.isEmpty {
                String(localized: "Wyzie request failed (HTTP \(code)): \(message)")
            } else {
                String(localized: "Wyzie request failed (HTTP \(code)).")
            }
        case .emptyFile:
            String(localized: "The downloaded subtitle file was empty.")
        case .fileTooLarge:
            String(localized: "The downloaded subtitle file was too large.")
        case .badURL:
            String(localized: "The subtitle URL was invalid.")
        }
    }
}

struct WyzieSubtitleQuery: Equatable, Sendable {
    let id: String
    let season: Int?
    let episode: Int?
    var language: String?
    var format: String?
    var hearingImpaired: Bool?
}

struct WyzieSubtitleService {
    static let shared = WyzieSubtitleService()

    private static let baseURL = "https://sub.wyzie.io"
    private static let maximumDownloadSize = 5 * 1024 * 1024

    private init() {}

    func search(_ query: WyzieSubtitleQuery, key: String) async throws -> [WyzieSubtitle] {
        let key = key.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { throw WyzieError.missingKey }

        var components = URLComponents(string: "\(Self.baseURL)/search")
        components?.queryItems = Self.queryItems(for: query, key: key)
        guard let url = components?.url else { throw WyzieError.badURL }

        let (data, response) = try await URLSession.shared.data(from: url)
        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 200
        return try Self.decodeSearchResponse(data, statusCode: statusCode)
    }

    func download(_ subtitle: WyzieSubtitle) async throws -> URL {
        guard let remoteURL = URL(string: subtitle.url),
              ["http", "https"].contains(remoteURL.scheme?.lowercased() ?? "")
        else {
            throw WyzieError.badURL
        }
        let (data, response) = try await URLSession.shared.data(from: remoteURL)

        if let http = response as? HTTPURLResponse,
           !(200..<300).contains(http.statusCode) {
            throw Self.statusError(data: data, statusCode: http.statusCode)
        }
        guard !data.isEmpty else { throw WyzieError.emptyFile }
        guard data.count <= Self.maximumDownloadSize else { throw WyzieError.fileTooLarge }

        guard let caches = FileManager.default.urls(
            for: .cachesDirectory,
            in: .userDomainMask
        ).first else {
            throw WyzieError.badURL
        }
        let directory = caches.appendingPathComponent("Subtitles", isDirectory: true)
        try FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true
        )
        let localURL = directory.appendingPathComponent(Self.cacheFileName(for: subtitle))
        try data.write(to: localURL, options: .atomic)
        return localURL
    }

    static func queryItems(for query: WyzieSubtitleQuery, key: String) -> [URLQueryItem] {
        var items = [URLQueryItem(name: "id", value: query.id)]
        if let season = query.season {
            items.append(URLQueryItem(name: "season", value: String(season)))
            if let episode = query.episode {
                items.append(URLQueryItem(name: "episode", value: String(episode)))
            }
        }
        if let language = query.language {
            items.append(URLQueryItem(name: "language", value: language))
        }
        if let format = query.format {
            items.append(URLQueryItem(name: "format", value: format))
        }
        if let hearingImpaired = query.hearingImpaired {
            items.append(URLQueryItem(name: "hi", value: String(hearingImpaired)))
        }
        items.append(URLQueryItem(name: "key", value: key))
        return items
    }

    static func cacheFileName(for subtitle: WyzieSubtitle) -> String {
        let id = sanitized(subtitle.id)
        let language = sanitized(subtitle.language)
        let sanitizedFormat = sanitized(subtitle.format)
        let format = sanitizedFormat.isEmpty ? "srt" : sanitizedFormat
        return "wyzie-\(id)-\(language).\(format)"
    }

    static func decodeSearchResponse(
        _ data: Data,
        statusCode: Int
    ) throws -> [WyzieSubtitle] {
        guard (200..<300).contains(statusCode) else {
            throw statusError(data: data, statusCode: statusCode)
        }
        return try JSONDecoder().decode([WyzieSubtitle].self, from: data)
    }

    private static func statusError(data: Data, statusCode: Int) -> WyzieError {
        struct ErrorEnvelope: Decodable {
            let code: Int
            let message: String
            let details: String?
            let notice: String?
        }

        let message = try? JSONDecoder().decode(ErrorEnvelope.self, from: data).message
        return .badStatus(code: statusCode, message: message)
    }

    private static func sanitized(_ value: String) -> String {
        value.lowercased().unicodeScalars
            .filter {
                CharacterSet.alphanumerics.contains($0)
                    || $0 == "-"
                    || $0 == "_"
            }
            .map(String.init)
            .joined()
    }
}
